jsonschema/tests/test_cli.py

```diff
@@ -66,36 +67,118 @@ def test_find_validator_in_jsonschema(self):
         )
         self.assertIs(arguments["validator"], Draft4Validator)
 
+    def test_none_validator(self):
+        arguments = cli.parse_args(
+            [
+                "--instance", self.instance_file,
+                self.schema_file,
+            ]
+        )
+        self.assertIs(arguments["validator"], LatestValidator)
+
+    def test_none_instance(self):
+        arguments = cli.parse_args(
+            [
+                self.schema_file,
+            ]
+        )
+        self.assertEqual(arguments["instances"], [])
+
+    def test_unknown_output(self):
+        # Avoid the help message on stdout
+        argparse_mock = MagicMock()
+        with patch('argparse.ArgumentParser._print_message', argparse_mock):
+            with self.assertRaises(SystemExit):
+                arguments = cli.parse_args(
+                    [
+                        "--output", "foo",
+                        self.schema_file
+                    ]
+                )
+
+    def test_useless_error_format(self):
+        # Avoid the help message on stdout
+        argparse_mock = MagicMock()
```

Comment:
This looks suspicious, definitely would rather well, not use mock at all, but if we're forced into it (which sometimes happens with argparse :/) then not be patching private things.

I don't see you asserting about stdout/error here, what happened if you remove this that wasn't desirable? In theory I'd just expect this test to assert that we exit-nonzero for this case, even when the instance and schema are valid.

Reply:
I’m sorry I don’t get it.

`test_unknown_output` and `test_useless_error_format` are checking cases of invalid options given to `argparse`. In command line, argparse stops the program and returns a non-zero code. But here I cannot test the command line, I therefore directly call `cli.parse_args` which itself calls `argparse.parse_args` which **raises an exception** and **writes on stderr** when incorrect options are given. The consistency check of `--output` and `--error-format` is manually done by us, but an exception is raised in the same way thanks to `parser.error()`.

Hence the mock and this test. I don’t know how to capture *stderr* otherwise, and the exception seems to be the only indicator of a failure. If I was able to capture *stderr*, I agree we could also check that it is not empty.

Am I missing something ?

Reply:
About the unittest mock, I found another way to capture *stderr*, by creating a custom `@contextmanager` (in `_helpers.py`).

Reply:
And finally, since I can capture these output, I added a test on both of them. Nothing must be in `stdout`, and a specific string must be in `stderr`. Maybe that is closer to what you expected.